apps/libsignpost/port_signpost_arduino.cpp

```diff
@@ -0,0 +1,252 @@
+#include <stdarg.h>
+#include <Arduino.h>
+#include <HardwareSerial.h>
+#include <Wire.h>
+#include "signbus_io_interface.h"
+#include "port_signpost.h"
+#include "signpost_entropy.h"
+
+//Defines for Arduino pins
+//Current values are defined for the Arduino MKRZero
+#define ARDUINO_MOD_IN 			4
+#define ARDUINO_MOD_OUT 		5
+#define ARDUINO_PPS				3 //placeholder pin
+#define DEBUG_LED 				LED_BUILTIN
+
+#define SIGNPOST_I2C_SPEED 		400000
+#define _PRINTF_BUFFER_LENGTH_	512
+
+//enable/disable printing debug messages over uart
+#define ENABLE_ARDUINO_PORT_PRINTF
+
+//TODO: Add error handling and error codes
+//Callback helper functions
+//Function protypes for Arduino callbacks and signpost callbacks differ,
+//so helper functions are used as an interface between the two.
+//The arduino hardware interrupt calls the helper function, which then calls the 
+//signpost callback.
+static void mod_in_callback_falling_helper();
+static void mod_in_callback_rising_helper();
+static void slave_listen_callback_helper(int num_bytes);
+static void slave_read_callback();
+
+//TODO: Implement separate i2c bus solely for use with signpost.
+
+//Make sure to set i2c buffer size to 256 within hardware i2c library
+//For a regular avr arduino, this is located in twi.h
+//For the arduino MKRZero, this is located in RingBuffer.h
+//This should be done automatically when building the Arduino hardware package
+
+//Buffer for port_printf
+static char g_port_printf_buf[_PRINTF_BUFFER_LENGTH_];
+//i2c address of arduino module
+static uint8_t g_arduino_i2c_address;
+//Callbacks for port layer functions
+static port_signpost_callback g_mod_in_callback_falling = NULL;
+static port_signpost_callback g_mod_in_callback_rising = NULL;
+static port_signpost_callback g_slave_listen_callback = NULL;
+//Slave receive buffer
+static uint8_t* g_slave_receive_buf = NULL;
+static size_t g_slave_receive_buf_max_len = 0;
+static size_t g_slave_receive_buf_len = 0;
+//Slave transmission buffer
+static uint8_t* g_slave_transmit_buf = NULL;
+static size_t g_slave_transmit_buf_len = 0;
+
+int port_signpost_init(uint8_t i2c_address) {
+	//Initialize i2c as slave
+	Wire.begin(i2c_address);
+	g_arduino_i2c_address = i2c_address;
+	//Setup GPIO pins
+	pinMode(ARDUINO_MOD_IN, INPUT);
+	pinMode(ARDUINO_MOD_OUT, OUTPUT);
+	pinMode(ARDUINO_PPS, INPUT);
+	pinMode(DEBUG_LED, OUTPUT);
+	//Set callback helpers for receiving and transmitting i2c message in slave mode
+	Wire.onReceive(slave_listen_callback_helper);
+	Wire.onRequest(slave_read_callback);
+	return 0;
+}
+
+int port_signpost_i2c_master_write(uint8_t addr, uint8_t* buf, size_t len) {
+	if (len == 0) {
+		return -1;
+	}
+	//Initialize i2c as master
+	Wire.begin();
+	//Set i2c clock speed
+	Wire.setClock(SIGNPOST_I2C_SPEED);
+	//Send i2c message to addr
+	Wire.beginTransmission(addr);
+	int num_written = Wire.write(buf, len);
+	int rc = Wire.endTransmission();
```

Comment:
I remember that I had some issues when I returned those return codes with port_signpost_i2c_master_write, so I just stopped using the codes. I kept them there in case we want to use them in the future. https://www.arduino.cc/en/Reference/WireEndTransmission